Refactor Trainers to use Policy

awjuliani opened the pull request:
* Introduces new abstraction of `Policy` between `Model` and `Trainer`. 
   * `Model` now contains all TensorFlow graph code.
   * `Policy` contains code to process calls to `Model` nodes.
   * `Trainer` contains high-level logic for collecting/processing experiences and calling `Policy` when needed.
* `Policy` uses dictionaries to keep track of valid inference and update calls. This prevents `Trainer` from needing to know anything about the specifics of the model.
* `Policy` keeps track of model meta-data such as whether visual observations are used.
* `Trainer` no longer un-necessarily takes references to environment.

Tested on: BananaCollector, BananaIL, Pyramid, and Hallway scenes.

Todo:
- [x] Add buffer -> batch function.
- [x] Update tests.

vincentpierre commented on python/unitytrainers/bc/trainer.py:
```diff
@@ -139,21 +109,7 @@ def take_action(self, all_brain_info: AllBrainInfo):
-            return agent_action, memories, None, None, None
-        else:
-            agent_action = self.sess.run(self.inference_run_list, feed_dict)
+        agent_action = self.policy.evaluate(agent_brain)
```
What if uses recurrent ?

mmattar commented:
@vincentpierre / @awjuliani - great PR - a lot of good changes. Happy to do a full review - just add me to the reviewer list when ready.

One high-level comment: we can remove all tf dependencies from the Policy class and push that to the Model class (we can possibly use the Factory Pattern to create a model instance, if necessary).

awjuliani commented on python/unitytrainers/buffer.py:
```diff
@@ -169,6 +170,18 @@ def shuffle(self, key_list=None):
             for key in key_list:
                 self[key][:] = [self[key][i] for i in s]
 
+        def make_mini_batch(self, start, end):
```
Added.

awjuliani commented on python/unitytrainers/policy.py:
```diff
@@ -0,0 +1,94 @@
+        :param brain_info: BrainInfo input to network.
+        :return: Output from policy based on self.inference_dict.
+        """
+        return None
```
Added.

awjuliani commented on python/unitytrainers/policy.py:
```diff
@@ -0,0 +1,94 @@
+        :param mini_batch: Batch of experiences.
+        :return: Results of update.
+        """
+        return None
```
Added.

awjuliani commented on python/unitytrainers/trainer_controller.py:
```diff
@@ -72,8 +72,8 @@ def __init__(self, env_path, run_id, save_freq, curriculum_folder,
-                    docker_target_name=docker_target_name,
-                    curriculum_folder=curriculum_folder)
+                        docker_target_name=docker_target_name,
+                        curriculum_file=curriculum_folder)
```
Fixed.

awjuliani commented on python/unitytrainers/ppo/policy.py:
```diff
@@ -0,0 +1,210 @@
+        :param mini_batch: Experience batch.
+        :return: Output from update process.
+        """
+        feed_dict = {self.model.batch_size: num_sequences,
```
It isn't clear to me that the approach you are describing is any more elegant than the current version.

awjuliani commented on python/unitytrainers/ppo/trainer.py:
```diff
@@ -169,30 +112,15 @@ def take_action(self, all_brain_info: AllBrainInfo):
-            return run_out[self.model.output], None, None, run_out[self.model.value], run_out
+            return run_out['action'], None, None, run_out['value'], run_out
 
     def construct_curr_info(self, next_info: BrainInfo) -> BrainInfo:
```
It current requires access to a few too many things as of right now. Definitely in the future should be refactored, but out of the scope.

awjuliani commented on python/unitytrainers/trainer_controller.py:
```diff
@@ -140,17 +140,17 @@ def _process_graph(self):
                 if self.trainers[brain_name].parameters["trainer"] \
-                   == "imitation":
+                        == "imitation":
                     nodes += [scope + x for x in ["action"]]
```
It currently exposes a few un-necessary nodes.

awjuliani commented:
@vincentpierre I've addressed all your comments.

@mmattar Thanks for the offer. I will add you now. I like the factory idea, and will add it. Feel free to start a review when you get a chance though.

mmattar commented on python/unitytrainers/policy.py:
```diff
@@ -0,0 +1,117 @@
+
+class Policy(object):
+    """
+    Contains a TensorFlow model, and the necessary
```
Does Policy need to assume that the model is a TF one?

mmattar commented on python/unitytrainers/policy.py:
```diff
@@ -0,0 +1,117 @@
+    pass
+
+
+class Policy(object):
```
I really like the Policy abstraction! 

This comment is a high-level one about PPOPolicy and BCPolicy: they both contain feed_dict manipulations that are: 1. somewhat hard to parse, particularly as they appear in sequence, 2. common across both classes, 3. Use string literals (e.g.  'prev_action') that should really be class constants.

I think it would be much easier to parse if we had a class that was responsible for generating a feed_dict object and that class would leverage a series of static helper methods that update a feed_dict, the common ones can live in Policy and the algorithm-specific ones can live in their respective classes.

Listing sample common ones I've noticed, but there could be more:

```
for i, _ in enumerate(brain_info.visual_observations):
            feed_dict[self.model.visual_in[i]] = brain_info.visual_observations[i]
```

```
if self.use_vec_obs:
            feed_dict[self.model.vector_in] = brain_info.vector_observations
```

```
if not self.use_continuous_act:
            feed_dict[self.model.action_masks] = brain_info.action_masks
```

mmattar commented on python/unitytrainers/ppo/policy.py:
```diff
@@ -0,0 +1,218 @@
+                    [-1, self.vec_obs_size])
+        if self.model.vis_obs_size > 0:
+            for i, _ in enumerate(self.model.visual_in):
+                _obs = mini_batch['visual_obs%d' % i]
```
Its `visual_obs` here, but `visual_observations` in BCPolicy.

mmattar commented on python/unitytrainers/bc/trainer.py:
```diff
@@ -237,7 +195,7 @@ def process_experiences(self, current_info: AllBrainInfo, next_info: AllBrainInf
                 agent_id = info_teacher.agents[l]
                 self.training_buffer.append_update_buffer(agent_id, batch_size=None,
-                                                          training_length=self.sequence_length)
+                                                          training_length=self.policy.sequence_length)
```
Either

```
self.training_buffer.append_update_buffer(
    agent_id, batch_size=None, training_length=self.policy.sequence_length)
```
or
```
self.training_buffer.append_update_buffer(agent_id, 
                                          batch_size=None, 
                                          training_length=self.policy.sequence_length)
```

awjuliani commented:
Thanks for the feedback, @mmattar. I've addressed all your comments (to a certain extent). The one thing I haven't done yet is to replace the use of strings as references. This is tied to a bigger refactor that would need to be done to how data is stored and organized in the buffer, and outside the scope of this PR. I also agree about the need to have the feed_dict generation as clean as possible. I've added a couple shared helper functions which propagate it. The issue is that exceptions are more the rule than rules, and most of the code cannot be shared between different models and update/evaluate. What can be shared though I made a point of sharing.
